### PR TITLE
WIP: download and run BLAST

### PR DESCRIPTION
**Pull request by jeremyadamsfisher**

### CODE REVIEW

File: Snakefile [hunk 1]
```diff
@@ -37,11 +40,30 @@ rule convert_sam_to_fastq:
...
+    input:
+        FTP.remote("https://ftp.ncbi.nlm.nih.gov/blast/db/ref_viruses_rep_genomes.tar.gz", keep_local=True)
+    output:
+        directory("./input/blastdb")
+    run:
+        "tar -xzvf {input} -C {output}"
+
+rule blast_against_known_viruses
+    input:
+        directory("./output/contigs/{sample}")  # all assembled contigs not matching the human genome as a fasta
```
Comment (jeremyadamsfisher) [inline Snakefile]: Can't use a directory as an input (idk why). But you can use:
```
rules.contig_assembly.output[0]
```

File: Snakefile [hunk 2]
```diff
@@ -37,11 +40,30 @@ rule convert_sam_to_fastq:
...
+    output:
+        "./output/viral_blast_results"
+    shell:
+        blastn -remote -db ./input/blastdb/ -outfmt 5 -query {input} -out {output}
+        # conda install -c bioconda blastn
```
Comment (jeremyadamsfisher) [inline Snakefile]: You'll need a conda environment spec like `bioinformatics-project/workflow/envs/spades.yaml `

Comment (jeremyadamsfisher) [reply Snakefile]: Perhaps instead of `./input/blastdb/`, use `rules.download_viral_blastdb.output[0]`

**Comment by DJSiddharthVader** [comment]
I was testing using blastn with conda and it does not appear to work, it raises this error
```blastn: error while loading shared libraries: libnsl.so.1: cannot open shared object file: No such file or directory```
 which appears to be Fedora/CentOS specific. The conda recipe relies on `libnsl1` but Fedora/CentOS dropped `libnsl1` for `libnsl2`, as per  [here](https://github.com/conda-forge/cmake-feedstock/issues/69), so it could be fixed with a simple simlink as stated [here](https://github.com/oracle/node-oracledb/issues/892#issuecomment-387082011) but obviously we dont have root permissions on the cluster. If you try this on a Debian based system if should work out of the box as they will still have `libnsl1` but bridges2 (and most clusters) are CentOS so ehh. 

Bridges2 does provide its own blast install, using `module load BLAST` which works fine but is outside of conda. Not sure what to do but for the sake of producing results we can just rely on the Bridges2 blast install which seems to work fine.

**Review by DJSiddharthVader — APPROVED** [review]
making the blastdb works and blastn should work on the bridges2 cluster but not tested in snakemake (blastn works on the CLI using the briges blast module)

**Comment by jeremyadamsfisher** [comment]
That libnsl1 issue is pretty rough. I appreciate you finding a workaround. Lets do this for the moment: make a very loud note in the manual/readme that `blastn` cannot be installed by conda in CentOS and that we require it to be installed by the user. Then, we can have an example installation like:
> If using a HPC, you may be able to load a version of `blastn` by loading a module like so: `module load blast`

**Comment by jeremyadamsfisher** [comment]
I installed `blast` via conda and didnt have get an issue:
```
(tmp) [jeremyf@bridges2-login013 ~]$ blastn
BLAST query/options error: Either a BLAST database or subject sequence(s) must be specified
Please refer to the BLAST+ user manual.
```
Where does the error come from? While using the query command?

**Comment by DJSiddharthVader** [comment]
Can you enter `$ which blastn` to make sure its calling the conda install of the bridges2 `blastn`?
I just logged onto bridges2 and ran the following
```
$ module load conda
$ conda create -n blast_error
$ conda init bash
$ conda activate blast_error
$ conda install -c bioconda blast
$ which blastn
~/.conda/envs/blast_error/bin/blastn
$ blastn
blastn: error while loading shared libraries: libnsl.so.1: cannot open shared object file: No such file or directory
```

If you are using the conda install of blastn then I am not sure what the issue is. I also tried this in the interact shell on bridges2 and get the same error.

I cloned my own `.bashrc` onto bridges so maybe that is causing some issue for me but I doubt it.

**Comment by jeremyadamsfisher** [comment]
Ok, I think the issue is that conda tries to install blast 2.5 by default, which throws that error. When I force it to use the latest version, it works fine:
```
(tmp) [jeremyf@bridges2-login013 bioinformatics-project]$ conda install -c bioconda blast
(tmp) [jeremyf@bridges2-login013 bioinformatics-project]$ blastn
blastn: error while loading shared libraries: libnsl.so.1: cannot open shared object file: No such file or directory
(tmp) [jeremyf@bridges2-login013 bioinformatics-project]$ conda install -c bioconda blast=2.10.1
(tmp) [jeremyf@bridges2-login013 bioinformatics-project]$ blastn
BLAST query/options error: Either a BLAST database or subject sequence(s) must be specified
```

**Comment by jeremyadamsfisher** [comment]
I fixed the above issue in another branch thats based off a major refactor, so I'm going to open a fresh PR

**Pull request closed without merging**